ojdbc-provider-jackson-oson/src/main/java/oracle/jdbc/provider/oson/OsonFactory.java

```diff
@@ -0,0 +1,402 @@
+  }
+
+  /**
+   * Creates a JSON parser from a Input Stream.
+   *
+   * @param in The inputStream containing JSON content.
+   * @return A custom OsonParser instance.
+   */
+  @Override
+  public JsonParser createParser(InputStream in) throws IOException {
+    IOContext ctxt = _createContext(in, true);
+    return _createParser(in, ctxt);
+  }
+
+  /**
+   * Creates a JSON parser from a Json String.
+   *
+   * @param content The String containing JSON content.
+   * @return A custom OsonParser instance.
+   */
+  @Override
+  public JsonParser createParser(String content) throws IOException {
+    IOContext ctxt = _createContext(content, true);
+    return _createParser(content.toCharArray(), 0, content.length(), ctxt, false);
+  }
+
+  /**
+   * Method for constructing JSON parser instance to parse contents of resource reference by given URL.
+   *
+   * @param url URL pointing to resource that contains JSON content to parse
+   *
+   * @return
+   * @throws IOException
+   */
+  @Override
+  public JsonParser createParser(URL url) throws IOException {
+    IOContext ctxt = _createContext(url, true);
+    InputStream in = _optimizedStreamFromURL(url);
+    if (_inputDecorator != null) {
+      in = _inputDecorator.decorate(ctxt, in);
+    }
+    return _createParser(in, ctxt);
+  }
+
+  /**
+   * Creates a JSON parser from a byte[]
+   *
+   * @param data Buffer that contains data to parse
+   * @param offset Offset of the first data byte within buffer
+   * @param len Length of contents to parse within buffer
+   *
+   * @return
+   * @throws IOException
+   */
+  @Override
+  public JsonParser createParser(byte[] data, int offset, int len) throws IOException {
+    IOContext ctxt = _createContext(data, true);
+    if (_inputDecorator != null) {
+      InputStream in = _inputDecorator.decorate(ctxt, data, offset, len);
+      if (in != null) {
+        return _createParser(in, ctxt);
+      }
+    }
+    return _createParser(data, offset, len, ctxt);
+  }
+
+  /**
+   * Creates a JSON parser from a char[].
+   *
+   * @param content char[] containing the JSON data
+   * @param offset offset of the first json data in char[]
+   * @param len length of the data
+   * @return
+   * @throws IOException
+   */
+  @Override
+  public JsonParser createParser(char[] content, int offset, int len) throws IOException {
+    IOContext ctxt = _createContext(content, true);
+    Reader r = new CharArrayReader(content, offset, len);
+    if (_inputDecorator != null) {
+      r = _inputDecorator.decorate(ctxt, r);
+    }
+    return _createParser(r, ctxt);
+  }
+
+
+  /**
+   * Creates a OSON parser from a InputStream
+   *
+   * @param in InputStream to use for reading content to parse
+   * @param ctxt I/O context to use for parsing
+   *
+   * @return
+   */
+  @Override
+  public JsonParser _createParser(InputStream in, IOContext ctxt) {
+    return new OsonParser(ctxt, _factoryFeatures, factory.createJsonBinaryParser(in));
```

Comment:
Why don't we use the decorator in this case?

Reply:
Fixed. Added the decorators if it were missing. We can't add decorator to the Reader, as we don't own it?

Reply:
You mean you can not add:

Reply:
Why?